crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1653,48 +1653,71 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_fstring_expression(&mut self, fstring: &ast::ExprFString) -> Type<'db> {
         let ast::ExprFString { range: _, value } = fstring;
 
+        let mut done = false;
+        let mut has_expression = false;
+        let mut concatenated = String::new();
         for part in value {
             match part {
-                ast::FStringPart::Literal(_) => {
-                    // TODO string literal type
+                ast::FStringPart::Literal(literal) => {
+                    concatenated.push_str(&literal.value);
                 }
                 ast::FStringPart::FString(fstring) => {
-                    let ast::FString {
-                        range: _,
-                        elements,
-                        flags: _,
-                    } = fstring;
-                    for element in elements {
-                        self.infer_fstring_element(element);
+                    for element in &fstring.elements {
+                        match element {
+                            ast::FStringElement::Expression(expression) => {
+                                let ast::FStringExpressionElement {
+                                    range: _,
+                                    expression,
+                                    debug_text: _,
+                                    conversion,
+                                    format_spec,
+                                } = expression;
+                                // Always infer sub-expressions, even if we've figured out the type
+                                let ty = self.infer_expression(expression);
+                                if !done {
+                                    // TODO: handle format specifiers by calling a method
+                                    // (`Type::format`?) that handles the `__format__` method.
+                                    // Conversion flags should be handled before calling
+                                    // `__format__`.
+                                    // https://docs.python.org/3/library/string.html#format-string-syntax
+                                    if !conversion.is_none() || format_spec.is_some() {
```

Comment:
nit: why `!conversion.is_none()` instead of `conversion.is_some()`?

Reply:
Because conversion is not an option but an enum with the `None` variant and the `Is` macro derived